### ISSUE TITLE
Adiabatic time limit returns nan

### ISSUE THREAD

**Issue by EFrion** [NumCosmo/NumCosmo#51]
Hi everyone,

I have an issue with finding the adiabatic time limit for the model I am using. When I print the function ```eval_adiab_at(None,-1)``` from the CSQ1D class, I find that ```alpha=1.9695613565250207e-09, dgamma=nan, alpha_reltol=2.1428272539007795e-27, dgamma_reltol=nan```, so ```dgamma``` seems to be the root of my problem. I don't know what these variables represent, so I am not sure how to approach this problem. Would you have some references I could look at which explain what this function does?

Many thanks!

**Comment by vitenti** [comment]
Hello @EFrion , this code is very experimental, we are trying new ways to define vacuum and we are testing new implementations. The problem is that some methods only accept positive time variables. Since your model seems to be symmetric I used the positive time branch and managed to get some results. I don't know the model you are computing so I have no idea if the results make sense. I also modified the solve_traj method since it was too slow to recompute all the time.

```
#!/usr/bin/python3

try:
  import gi
  gi.require_version('NumCosmo', '1.0')
  gi.require_version('NumCosmoMath', '1.0')
except:
  pass
  
from gi.repository import NumCosmo as Nc
from gi.repository import NumCosmoMath as Ncm

import math
import sys
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from scipy.integrate import odeint
from scipy.interpolate import interp1d
from numpy import diff,gradient



###
### Helper function to write latex formatted numbers
###

def latex_float(f):
	float_str = "{0:.2g}".format(f)
	if "e" in float_str:
		base, exponent = float_str.split("e")
		if base == 1.0:
			return r"10^{{{1}}}".format (int(exponent))
		else:
			return r"{0} \times 10^{{{1}}}".format(base, int(exponent))
	else:
		return float_str

###
### Initialising the library
###

__name__ = "NcContext"
Ncm.cfg_init ()
Ncm.cfg_set_log_handler (lambda msg: sys.stdout.write (msg) and sys.stdout.flush ())



###
### Define Class
###

class PyCSQ1DMagBounceExpPot (Ncm.CSQ1D):
	def __init__ (self, s2 = 81.0, dd = -0.0009, bb = 1.0, h = 0.7, beta = 1.0, lp = 1.0): 
	
		Ncm.CSQ1D.__init__(self)
		
		self.h = h # Hubble rate
		self.lp = lp # Planck length
		self.s2 = s2 # \sigma^2
		self.dd = dd # 
		self.bb = bb # Strength Gaussian coupling
		self.beta = beta # Width Gaussian coupling
		self.RH = (Ncm.C.c () / (1.0e5 * h)) # Hubble radius in units of Mpc
		self.solved_traj = None
		

	### Bohmian trajectories alpha(t) and phi(t).
	def traj(self,U,t):
		a,p = U
		
		s2 = self.s2
		dd = self.dd
		lp = self.lp
		T1 = p * s2 * math.sin(2.0 * dd * a)
		T2 = 2.0 * dd * math.sinh(s2 * a * p)
		T3 = 2.0 * lp * math.exp(3.0 * a)
		T4 = math.cos(2.0* dd * a)
		T5 = math.cosh(s2 * a * p)
		T6 = -a * s2 * math.sin(2.0 * dd * a)
		T7 = 2.0 * dd * math.cos(2.0 * dd * a)
		T8 = 2.0 * dd* math.cosh(s2 * a * p)
		dadt = (T1+T2)/(T3*(T4+T5))
		dpdt = (T6+T7+T8)/(T3*(T4+T5))
		return [dadt, dpdt]

	### Solve Bohmian trajectories. Returns the scale factor a(t), alpha(t), phi(t), and their derivatives
	def solve_traj(self):
		# IC at the bounce
		
		if self.solved_traj:
		    return self.solved_traj
		
		a0 = 0.083163
		phi0 = 0.0

		# Cosmic Time
		tf = 2500.0
		ti = -tf
		npoints=10000
		t_after = np.linspace(0,tf,npoints) # Solve from 0 to tf
		t_before = np.linspace(0,ti,npoints) # Solve from 0 to -tf
		tphi = np.linspace(ti,tf,2*npoints) # Join the two solutions

		# Solving the two ODEs. Returns a vector containing alpha(t) and phi(t)
		sol_after = odeint(self.traj,[a0,phi0],t_after)
		sol_before = odeint(self.traj,[a0,phi0],t_before)
		
		# Separate alpha(t) and phi(t), and construct the whole solutions from -tf to tf
		alpha_after = sol_after[:,0]
		alpha_before = sol_before[:,0]
		alpha_before = alpha_before[::-1]
		alpha_tot = np.append(alpha_before,alpha_after)
		
		phi_after = sol_after[:,1]
		phi_before = sol_before[:,1]
		phi_before = phi_before[::-1]
		phi_tot = np.append(phi_before,phi_after)
		
		# Get scale factor a(t) = exp(alpha(t))
		a_t_after = np.exp(sol_after[:,0])
		a_t_before = np.exp(sol_before[:,0])
		a_t_before = a_t_before[::-1]
		a_tot = np.append(a_t_before,a_t_after)

		# Interpolate the result to make the functions callable. This is needed to use the C functions from csq1d.c
		a_t = interp1d(tphi,a_tot,axis=0,fill_value="extrapolate") # Extrapolate should be fine at large |t|, since the functions are smooth.
		alpha_t = interp1d(tphi,alpha_tot,axis=0,fill_value="extrapolate")
		phi_t = interp1d(tphi,phi_tot,axis=0,fill_value="extrapolate")
		
		
		# Store the derivative of alpha and phi with cosmic time. These are the same definitions as in the the traj() function
		alphadot = []
		phidot = []
		s2 = self.s2
		dd = self.dd
		lp = self.lp
		for t in range(len(tphi)):
			T1 = phi_tot[t] * s2 * math.sin(2.0 * dd * alpha_tot[t])
			T2 = 2.0 * dd * math.sinh(s2 * alpha_tot[t] * phi_tot[t])
			T3 = 2.0 * lp * math.exp(3.0 * alpha_tot[t])
			T4 = math.cos(2.0* dd * alpha_tot[t])
			T5 = math.cosh(s2 * alpha_tot[t] * phi_tot[t])
			T6 = -alpha_tot[t] * s2 * math.sin(2.0 * dd * alpha_tot[t])
			T7 = 2.0 * dd * math.cos(2.0 * dd * alpha_tot[t])
			T8 = 2.0 * dd* math.cosh(s2 * alpha_tot[t] * phi_tot[t])
			alphadot.append((T1+T2)/(T3*(T4+T5)))
			phidot.append((T6+T7+T8)/(T3*(T4+T5)))
		
		alphadot_t = interp1d(tphi,alphadot,axis=0,fill_value="extrapolate")
		phidot_t = interp1d(tphi,phidot,axis=0,fill_value="extrapolate")

		
		# Store the second derivative of phi(t).
		phidotdot = []
		for t in range(len(tphi)):
			T1 = (3.0 * math.sin(4.0 * dd *alpha_tot[t]) - 4.0 * dd) * math.sinh(s2 * alpha_tot[t] * phi_tot[t])
			T2 = (3.0 * math.sin(2.0 * dd *alpha_tot[t]) - 2.0 * dd * math.cos(2.0 * dd * alpha_tot[t])) * math.sinh(2.0 * s2 * alpha_tot[t] * phi_tot[t])
			T3 = 2.0 * s2 * phi_tot[t] * math.sinh(s2 * alpha_tot[t] * phi_tot[t]) * (math.sin(2.0 * dd * alpha_tot[t]))**2
			T4 = (2.0 * dd *(math.cosh(2.0* s2 * alpha_tot[t] * phi_tot[t])-3.0) + 3.0 * math.sin(4.0 * dd * alpha_tot[t])) * math.sin(2.0 * dd * alpha_tot[t])
			T5 = (-2.0 * dd * math.sin(4.0 * dd * alpha_tot[t]) - 3.0 * math.cos(4.0 * dd *alpha_tot[t]) + 3.0) * math.cosh(s2 * alpha_tot[t] * phi_tot[t])
			T6 = s2 * phi_tot[t] * math.sin(2.0 * dd * alpha_tot[t]) + 2.0 * dd * math.sinh(s2 * alpha_tot[t] * phi_tot[t])
			T7 = math.cos(2.0 * dd *alpha_tot[t]) + math.cosh(s2 * alpha_tot[t] * phi_tot[t])
			T8 = 6.0 * dd * math.cos(2.0 * dd *alpha_tot[t]) + 6.0 * dd * math.cosh(s2 * alpha_tot[t] * phi_tot[t]) + s2 * math.sin(2.0 * dd * alpha_tot[t])
			T9 = (s2**3) * (alpha_tot[t]**3) * (math.cos(4.0 * dd * alpha_tot[t]) -1.0) * math.sinh(s2 * alpha_tot[t] * phi_tot[t])
			T10 = 2.0 * dd *(s2**2) * (alpha_tot[t]**2) * (math.sin(4.0 * dd * alpha_tot[t]) * math.sinh(s2 * alpha_tot[t] * phi_tot[t]) + math.sin(2.0 * dd * alpha_tot[t]) * math.sinh(2.0 * s2 * alpha_tot[t] * phi_tot[t]))
			T11 = (lp**2) * 8.0 * math.exp(6.0 * alpha_tot[t]) * (math.cos(2.0 * dd *alpha_tot[t]) + math.cosh(s2 * alpha_tot[t] * phi_tot[t]))**3
			phidotdot.append((s2*alpha_tot[t]*(2.0 * dd * (T1+T2) + s2 * phi_tot[t] * (T3+T4) + T5) -2.0 * T6*T7*T8 + T9 + T10)/T11)
			
		phidotdot_t = interp1d(tphi,phidotdot,axis=0,fill_value="extrapolate")
		
		self.solved_traj = [a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t]
			
		return self.solved_traj


	### Gaussian Coupling. f^2(phi)=1/4 + B*e^(-phi^2/beta^2)
	def F_t(self,t):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		beta2= beta**2
		return np.sqrt(0.25 * (1.0 + 4.0*bb*np.exp(-(phi_t(t))**2 / beta2)))
		
	### Gaussian coupling to the square	
	def F2(self,t):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		beta2 = beta * beta
		return 0.25 * (1.0 + 4.0*bb*np.exp(-(phi_t(t))**2 / beta2))

	### f'(t)
	def fdot(self,t):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		beta2= beta**2
		return -bb * phi_t(t) * phidot_t(t) * np.exp(-phi_t(t)**2 / beta2) / beta2 / (self.F_t(t))
		
	### f'(t)/f(t)
	def fdotoverf(self,t):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		beta2= beta**2
		return -bb * phi_t(t) * phidot_t(t) * np.exp(-phi_t(t)**2 / beta2) / beta2 / (self.F2(t))
		
	### f''(t)
	def fdotdot(self,t):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		beta2= beta**2
		T1 = 2.0 * (phi_t(t)**2) * (phidot_t(t)**2) / ((self.F_t(t)) * beta2)
		T2 = phi_t(t) * phidot_t(t) * self.fdotoverf(t) / (self.F_t(t)) 
		T3 = phi_t(t) * phidotdot_t(t) / (self.F_t(t)) 
		T4 = (phidot_t(t)**2) / (self.F_t(t))
		return bb * np.exp(-phi_t(t)**2 / beta2) * (T1+T2+T3+T4) / beta2
		
	### f''(t)/f
	def fdotdotoverf(self,t):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		beta2= beta**2
		T1 = 2.0 * (phi_t(t)**2) * (phidot_t(t)**2) / ((self.F2(t)) * beta2)
		T2 = phi_t(t) * phidot_t(t) * self.fdotoverf(t) / (self.F2(t)) 
		T3 = phi_t(t) * phidotdot_t(t) / (self.F2(t)) 
		T4 = (phidot_t(t)**2) / (self.F2(t))
		return bb * np.exp(-phi_t(t)**2 / beta2) * (T1+T2+T3+T4) / beta2
		
	### Time when coupling is equal to 1 (i.e. background suffers no influence from EM field)	
	def Fconst_time(self):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		bb = self.bb
		beta = self.beta
		str(phi_t(1.0e3))
		return phi_t(beta *math.sqrt(-1.0* math.log(0.75 / bb)))


	### xi = ln(mn)
	def do_eval_xi(self,model,t, k):
		return np.log (k) + np.log (self.F_t (t))

	###	Frequency nu = k / a(t)
	def do_eval_nu(self,model,t, k):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		return k / a_t (t)
		
	### nu**2
	def do_eval_nu2(self,model,t, k):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		return (k / a_t (t))**2

	### Mass m = a(t) * f(t)
	def do_eval_m(self,model,t, k):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		return a_t(t) * self.F_t (t)


	### Adiabatic vacuum condition F_n<<F_n+1<<...<<F_1<<1
	### F_1 = [a(t)/2K] * [f'(t)/f(t)]
	def do_eval_F1(self,model,t, k):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		return  a_t (t) * self.fdotoverf (t) / (2.0*k)
	
	### F_2 = [a(t)/2k]^2 * [alpha'(t) * f'(t)/f(t) + f''(t)/f(t) - (f'(t)/f(t))^2]
	def do_eval_F2(self,model,t, k):
		[a_t,alpha_t,phi_t,alphadot_t,phidot_t,phidotdot_t] = self.solve_traj()
		return (a_t(t)/(2.0*k))**2 * (alphadot_t(t) * self.fdotoverf(t) + self.fdotdotoverf(t) - (self.fdotoverf(t)**2))
		
	### Needed in csq1d.c		
	def do_prepare(self,model):
		pass


###
### Initialise the model
###
csq1d = PyCSQ1DMagBounceExpPot()

### Time. For plotting purposes only. tf here must match tf in solve_traj()
tf = 2500.0
ti = -tf
npoints=10000
tphi = np.linspace(ti,tf,2*npoints)

[a_t,alpha,phi,alphadot,phidot,phidotdot] = csq1d.solve_traj()

####################################################################
### Mosaic plots for alpha, phi, and derivatives
 
fig = plt.figure(constrained_layout=True)
subfigs = fig.subfigures(1, 3, wspace=0.07)

ax0 = subfigs[0].subplots(2, 1)
ax0[0].plot(tphi,alpha(tphi),label=r'$\alpha(t)$')
ax0[0].legend(loc="best")
ax0[1].plot(tphi,phi(tphi),label=r'$\phi(t)$')
ax0[1].legend(loc="best")
subfigs[0].set_facecolor('0.9')
subfigs[0].suptitle('Trajectories')
subfigs[0].supxlabel('t')

ax1 = subfigs[1].subplots(2, 1)
ax1[0].plot(tphi,alphadot(tphi),label=r'$\dot{\alpha}(t)$')
ax1[0].legend(loc="best")
ax1[1].plot(tphi,phidot(tphi),label=r'$\dot{\phi}(t)$')
ax1[1].legend(loc="best")
subfigs[1].set_facecolor('0.9')
subfigs[1].suptitle('Derivatives')
subfigs[1].supxlabel('t')

ax2 = subfigs[2].subplots(2, 1)
ax2[0].plot(tphi,phidotdot(tphi),label=r'$\ddot{\phi}(t)$')
ax2[0].legend(loc="best")
ax2[1].plot(tphi,phi(tphi),label=r'$\phi(t)$')
ax2[1].set_xlim(-1.0,1.0)
ax2[1].set_ylim(-1.0e-15,1.0e-15)
subfigs[2].set_facecolor('0.9')
subfigs[2].suptitle('Second derivative')
subfigs[2].supxlabel('t')

plt.show()

####################################################################




####################################################################
### Mosaic plots for the coupling f and its derivatives
 
fig = plt.figure(constrained_layout=True)
subfigs = fig.subfigures(1, 3, wspace=0.07)

ax0 = subfigs[0].subplots(2, 1)
ax0[0].plot(tphi,csq1d.F_t(tphi),label=r'$f(t)$')
ax0[0].legend(loc="best")
ax0[1].plot(tphi,phi(tphi),label=r'$\phi(t)$')
ax0[1].legend(loc="best")
subfigs[0].set_facecolor('0.9')
subfigs[0].suptitle('Trajectories')
subfigs[0].supxlabel('t')

ax1 = subfigs[1].subplots(2, 1)
ax1[0].plot(tphi,csq1d.fdot(tphi),label=r'$\dot{f}(t)$')
ax1[0].legend(loc="best")
ax1[1].plot(tphi,csq1d.fdotoverf(tphi),label=r'$\dot{f}/f(t)$')
ax1[1].legend(loc="best")
subfigs[1].set_facecolor('0.9')
subfigs[1].suptitle('Derivatives')
subfigs[1].supxlabel('t')

ax2 = subfigs[2].subplots(2, 1)
ax2[0].plot(tphi,csq1d.fdotdot(tphi),label=r'$\ddot{f}(t)$')
ax2[0].legend(loc="best")
ax2[1].plot(tphi,csq1d.fdotdotoverf(tphi),label=r'$\ddot{f}/f(t)$')
ax2[1].legend(loc="best")
subfigs[2].set_facecolor('0.9')
subfigs[2].suptitle('Second derivative')
subfigs[2].supxlabel('t')

plt.show()

####################################################################



####################################################################
### Mosaic plots for the adiabatic time functions
 
fig = plt.figure(constrained_layout=True)
subfigs = fig.subfigures(1, 3, wspace=0.07)

ax0 = subfigs[0].subplots(2, 1)
ax0[0].plot(tphi,csq1d.do_eval_m(None,tphi,1.0),label=r'$m(t)$')
ax0[0].legend(loc="best")
ax0[1].plot(tphi,csq1d.do_eval_xi(None,tphi,1.0),label=r'$\xi(t)$')
ax0[1].legend(loc="best")
subfigs[0].set_facecolor('0.9')
subfigs[0].suptitle('Mass')
subfigs[0].supxlabel('t')

ax1 = subfigs[1].subplots(2, 1)
ax1[0].plot(tphi,csq1d.do_eval_nu(None,tphi,1.0),label=r'$\nu(t)$')
ax1[0].legend(loc="best")
ax1[1].plot(tphi,csq1d.do_eval_nu2(None,tphi,1.0),label=r'$\nu^2(t)$')
ax1[1].legend(loc="best")
subfigs[1].set_facecolor('0.9')
subfigs[1].suptitle('Frequency')
subfigs[1].supxlabel('t')

ax2 = subfigs[2].subplots(2, 1)
ax2[0].plot(tphi,csq1d.do_eval_F1(None,tphi,1.0),label=r'$F1(t)$')
ax2[0].legend(loc="best")
ax2[1].plot(tphi,csq1d.do_eval_F2(None,tphi,1.0),label=r'$F2(t)$')
ax2[1].legend(loc="best")
subfigs[2].set_facecolor('0.9')
subfigs[2].suptitle('Adiabatic Conditions')
subfigs[2].supxlabel('t')

plt.show()

####################################################################

### The ultimate goal is to find the adiabatic time limit, which fails.
### As tests, I used eval_at, which fails with a Segmentation fault at any time
### and eval_adiab_at, which returns a NaN for one reltol, and gives alpha = sinh(a_t) = 0 (different alpha than before)

csq1d.set_k (1.0e-2)
tAD, F1F1_min, t_Bl, t_Bu = csq1d.find_adiab_max (None, 1.0e1, 1.0e4, 1.0e-7)
tAD = 2500.0
csq1d.set_tf (tAD)
csq1d.set_init_cond_adiab (None, -tAD)
csq1d.prepare (None)
t_a, t_min = csq1d.get_time_array ()

phi_a = []
Pphi_a = []

for t in t_a:
    alpha, dgamma = csq1d.eval_at (t)
    phi, Pphi = csq1d.alpha_dgamma_to_phi_Pphi (None, t, alpha, dgamma)
    phi = phi[0] + 1j * phi[1]
    Pphi = Pphi[0] + 1j * Pphi[1]
    
    phi_a.append (phi)
    Pphi_a.append (Pphi)


fig = plt.figure(constrained_layout=True)

plt.plot (t_a, np.abs (phi_a))

plt.show()

fig = plt.figure(constrained_layout=True)

plt.plot (t_a, np.abs(np.real (phi_a)))
plt.plot (t_a, np.abs(np.imag (phi_a)))

plt.show()

fig = plt.figure(constrained_layout=True)

plt.plot (t_a, np.abs (Pphi_a))

plt.show()

fig = plt.figure(constrained_layout=True)

plt.plot (t_a, np.abs(np.real (Pphi_a)))
plt.plot (t_a, np.abs(np.imag (Pphi_a)))

plt.show()
```

**Comment by EFrion** [comment]
Thanks for the help!

I am surprised with the plots, though. I want to find the evolution of a gauge field and its momentum, similar to what we have done in [2004.07269](https://arxiv.org/abs/2004.07269), but this time using the background defined in [Bacalhau et al (2017)](https://arxiv.org/abs/1706.08830). Since the vacuum should be the same in this background, I would have expected a plot similar to the one in our paper, _i.e._, in which both the field and its momentum are constant at first, then start evolving.

I thought at first that it might have been a problem with my definitions, but I checked again and they have the expected behaviour.
Then, I changed the portion of code with the time loop to use the function ``csq1d.get_J_at`` instead of ``csq1d.alpha_dgamma_to_phi_Pphi`` to compare directly with what you did in our paper and check whether the two methods are equivalent or not, but I obtained the same results, so it seems that they are.

Any way, I close the issue since the code now returns something :)
